data_steward/retraction/create_deid_map.py

```diff
@@ -0,0 +1,172 @@
+retractions to obtain the correct research_id for the deid datasets.
+
+Create a module in the retraction package that generates all _deid_map tables for each release in the combined dataset
+if there is an associated deid_dataset. This will be created by querying the observation table in combined and joining
+to the observation table in deid on observation_id. Before the table creation, check to see if _deid_map or deid_map
+exists, if deid_map exists rename for consistency to _deid_map.
+"""
+
+# Python imports
+import re
+import argparse
+import logging
+
+# Third party imports
+import pandas as pd
+
+# Project imports
+import utils.bq
+import bq_utils
+from retraction.retract_utils import DEID_REGEX
+from constants.retraction import create_deid_map as consts
+
+
+def get_combined_datasets_for_deid_map(project_id):
+    """
+    List all datasets in given project_id, filter out datasets that contain a deid dataset and collect the corresponding
+    combined dataset
+    :param project_id: bq name of project_id
+    :return: list of combined_datasets that should contain a _deid_map table
+    """
+    all_datasets_obj = utils.bq.list_datasets(project_id)
+    all_datasets = [d.dataset_id for d in all_datasets_obj]
+    deid_datasets = []
+
+    # Keep only deid datasets with standard cdr release naming convention
+    release_datasets = [
+        d for d in all_datasets if consts.CURRENT_RELEASE_REGEX.match(d)
+    ]
+    # Filter for deid datasets
+    for dataset in release_datasets:
+        if bool(re.match(DEID_REGEX, dataset)) is True:
+            deid_datasets.append(dataset)
+    # Find corresponding combined dataset for each deid dataset
+    deid_and_combined_df = get_corresponding_combined_dataset(
+        all_datasets, deid_datasets)
+    return deid_and_combined_df
+
+
+def get_corresponding_combined_dataset(all_datasets, deid_datasets):
+    """
+    Takes in all datasets in a project and a list of all the deid datasets in that project and returns the corresponding
+    combined dataset to that deid dataset
+    :param all_datasets: list of all datasets in the given bq project
+    :param deid_datasets: list of all deid datasets in the given bq project
+    :return: list of combine datasets
+    """
+    deid_and_combined_datasets_df = pd.DataFrame(
+        columns=['deid_dataset', 'combined_dataset'])
+    # Get release from deid dataset for both older and current release tags and search for combined dataset
+    for d in deid_datasets:
+        release = d.split('_')[0]
+        prefix = release[1:]
+        combined = prefix + '_combined'
+
+        if combined in all_datasets:
+            new_row = pd.Series({
+                'deid_dataset': d,
+                'combined_dataset': combined
+            })
+            deid_and_combined_datasets_df = deid_and_combined_datasets_df.append(
+                new_row, ignore_index=True)
+
```

Comment:
It would be informative to have a log message here if a combined dataset is not found for a particular deid dataset.

Reply:
added!